File: Java/src/main/java/com/toaction/hot100/linkedList/SortList.java
package com.toaction.hot100.linkedList;

public class SortList {

    /*
    https://leetcode.cn/problems/sort-list/
    给你链表的头结点 head ，请将其按 升序 排列并返回 排序后的链表 。
    * */

    // 归并排序
    public static ListNode sortList(ListNode head) {
        if (head==null || head.next==null) {
            return head;
        }
        // 先找到中心节点
        ListNode mid = findMiddle(head);
        // 两部分分开
        ListNode head1 = head;
        ListNode head2 = mid.next;
        mid.next = null;
        // 排序
        head1 = sortList(head1);
        head2 = sortList(head2);
        // 合并
        return merge(head1, head2);
    }


    private static ListNode findMiddle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next!=null && fast.next.next!=null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    private static ListNode merge(ListNode head1, ListNode head2) {
        ListNode dump = new ListNode();
        ListNode cur = dump;
        while(head1!=null && head2!=null) {
            if (head1.val < head2.val) {
                cur.next = head1;
                head1 = head1.next;
            }else {
                cur.next = head2;
                head2 = head2.next;
            }
            cur = cur.next;
        }
        cur.next = head1==null?head2:head1;
        return dump.next;
    }
}

File: linkedlist/sort_list.go
package linkedlist

/*
148. 排序链表：https://leetcode.cn/problems/sort-list/

给你链表的头结点 head ，请将其按 升序 排列并返回 排序后的链表 。

示例：
	输入：head = [4,2,1,3]
	输出：[1,2,3,4]
*/

// 归并排序
func SortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	mid := findMiddleNode(head)
	next := mid.Next
	mid.Next = nil
	first, second := SortList(head), SortList(next)
	return MergeTwoLists(first, second)
}

// 返回链表中间节点（奇数）或前半部分的尾节点（偶数）
func findMiddleNode(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}
